File: tests/live-workout-contract.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import vm from "node:vm";

const sharedSource = await readFile("pwa/shared-workout.js", "utf8");
const liveSource = await readFile("pwa/live-workout.js", "utf8");

function contract() {
  const context = { TextEncoder };
  context.globalThis = context;
  vm.createContext(context);
  vm.runInContext(sharedSource, context);
  vm.runInContext(liveSource, context);
  return context.GymLiveWorkout;
}

const roomId = "lr_11111111111111111111111111111111";
const ownerProfile = {
  profileId: "p_11111111111111111111111111111111",
  displayName: "Owner"
};
const friendProfile = {
  profileId: "p_22222222222222222222222222222222",
  displayName: "Friend"
};

function livePlan() {
  return {
    version: 1,
    exercises: [{
      exerciseId: "e_01",
      catalogKey: "bench_press",
      name: "Bench Press",
      sets: [
        { setId: "s_01_01", weight: 80, reps: 8 },
        { setId: "s_01_02", weight: 82.5, reps: 6 }
      ]
    }]
  };
}

function planSummary() {
  return { exerciseCount: 1, setCount: 2, exerciseNames: ["Bench Press"] };
}

function activeSnapshot() {
  return {
    version: 1,
    room: {
      roomId,
      status: "active",
      roomRevision: 3,
      closeReason: null,
      createdAt: "2026-08-10T10:00:00Z",
      inviteExpiresAt: "2026-08-17T10:00:00Z",
      startedAt: "2026-08-10T10:05:00Z",
      activeExpiresAt: "2026-08-11T10:05:00Z",
      endedAt: null,
      summary: planSummary()
    },
    plan: livePlan(),
    participants: [{
      isSelf: true,
      profile: ownerProfile,
      role: "owner",
      state: "joined",
      membershipRevision: 1,
      joinedAt: "2026-08-10T10:00:00Z",
      finishedAt: null,
      departedAt: null,
      progress: {
        version: 1,
        revision: 2,
        completedSets: [{
          setId: "s_01_01",
          weight: 81,
          reps: 8,
          completedAt: "2026-08-10T10:06:00Z"
        }],
        undoableSetId: "s_01_01",
        finishedAt: null
      }
    }, {
      isSelf: false,
      profile: friendProfile,
      role: "participant",
      state: "joined",
      membershipRevision: 2,
      joinedAt: "2026-08-10T10:03:00Z",
      finishedAt: null,
      departedAt: null,
      progress: {
        version: 1,
        revision: 1,
        completedSets: [],
        undoableSetId: null,
        finishedAt: null
      }
    }]
  };
}

test("live snapshot preserves two independent progress lanes and canonical set ids", () => {
  const parsed = contract().snapshot(activeSnapshot());
  assert.equal(parsed.room.status, "active");
  assert.equal(parsed.participants.length, 2);
  assert.equal(parsed.participants[0].progress.completedSets[0].weight, 81);
  assert.equal(parsed.participants[1].progress.completedSets.length, 0);
  assert.deepEqual(JSON.parse(JSON.stringify(contract().sharedPlan(parsed.plan))), {
    version: 1,
    exercises: [{
      catalogKey: "bench_press",
      name: "Bench Press",
      sets: [{ weight: 80, reps: 8 }, { weight: 82.5, reps: 6 }]
    }]
  });
});

test("ready room accepts joined participants without progress before host start", () => {
  const value = activeSnapshot();
  value.room.status = "ready";
  value.room.roomRevision = 2;
  value.room.startedAt = null;
  value.room.activeExpiresAt = null;
  value.participants.forEach(participant => { participant.progress = null; });

  const parsed = contract().snapshot(value);
  assert.equal(parsed.room.status, "ready");
  assert.equal(parsed.participants.every(participant => participant.progress === null), true);
});

test("cancelled room may retain started timestamps for bounded recovery", () => {
  const value = activeSnapshot();
  value.room.status = "cancelled";
  value.room.closeReason = "left";
  value.room.endedAt = "2026-08-10T10:30:00Z";
  value.participants[1].state = "left";
  value.participants[1].departedAt = "2026-08-10T10:30:00Z";

  assert.equal(contract().snapshot(value).room.closeReason, "left");
});

test("contract rejects peer identity duplication, unknown fields and invalid calendar dates", () => {
  const duplicate = activeSnapshot();
  duplicate.participants[1].profile.profileId = ownerProfile.profileId;
  assert.throws(() => contract().snapshot(duplicate), /identities are inconsistent/);

  const extra = activeSnapshot();
  extra.room.note = "private";
  assert.throws(() => contract().snapshot(extra), /fields are invalid/);

  const invalidDate = activeSnapshot();
  invalidDate.room.createdAt = "2026-02-30T10:00:00Z";
  assert.throws(() => contract().snapshot(invalidDate), /timestamp is invalid/);
});

test("contract rejects duplicate and out-of-order canonical plan identities", () => {
  const duplicate = activeSnapshot();
  duplicate.plan.exercises.push({
    exerciseId: "e_02",
    name: "\u00a0bench press\u00a0",
    sets: [{ setId: "s_02_01", weight: 40, reps: 12 }]
  });
  duplicate.room.summary = {
    exerciseCount: 2,
    setCount: 3,
    exerciseNames: ["Bench Press", "\u00a0bench press\u00a0"]
  };
  assert.throws(() => contract().snapshot(duplicate), /identity is duplicated|summary is duplicated/);

  const order = activeSnapshot();
  order.plan.exercises[0].sets[0].setId = "s_01_02";
  assert.throws(() => contract().snapshot(order), /set is invalid/);
});

test("live inbox keeps invitations separate from joined rooms", () => {
  const parsed = contract().inbox({
    version: 1,
    invitations: [{
      roomId,
      status: "waiting",
      roomRevision: 1,
      createdAt: "2026-08-10T10:00:00Z",
      inviteExpiresAt: "2026-08-17T10:00:00Z",
      summary: planSummary(),
      owner: ownerProfile
    }],
    rooms: []
  });
  assert.equal(parsed.invitations[0].owner.displayName, "Owner");
  assert.equal(parsed.rooms.length, 0);

  assert.throws(() => contract().inbox({
    version: 1,
    invitations: [{
      roomId,
      status: "waiting",
      roomRevision: 1,
      createdAt: "2026-08-10T10:00:00Z",
      inviteExpiresAt: "2026-08-17T10:00:00Z",
      summary: planSummary(),
      owner: ownerProfile
    }],
    rooms: [{
      roomId,
      status: "ready",
      roomRevision: 2,
      role: "participant",
      memberState: "joined",
      membershipRevision: 2,
      createdAt: "2026-08-10T10:00:00Z",
      startedAt: null,
      activeExpiresAt: null,
      summary: planSummary(),
      peer: ownerProfile
    }]
  }), /duplicate rooms/);
});

test("mutation parsers bind result kind, ids, revisions and timestamps", () => {
  const api = contract();
  assert.equal(api.sendResult({
    version: 1,
    result: "submitted",
    roomId,
    status: "waiting",
    roomRevision: 1
  }).roomId, roomId);
  assert.equal(api.respondResult({
    version: 1,
    result: "joined",
    roomId,
    status: "ready",
    roomRevision: 2,
    membershipRevision: 2
  }).status, "ready");
  assert.equal(api.startResult({
    version: 1,
    result: "started",
    roomId,
    status: "active",
    roomRevision: 3,
    startedAt: "2026-08-10T10:05:00Z",
    activeExpiresAt: "2026-08-11T10:05:00Z",
    myProgressRevision: 1
  }).status, "active");
  assert.equal(api.applyResult({
    version: 1,
    result: "applied",
    roomId,
    roomRevision: 4,
    progressRevision: 2,
    kind: "complete_set",
    setId: "s_01_01",
    completedAt: "2026-08-10T10:06:00Z"
  }).progressRevision, 2);
  assert.equal(api.finishResult({
    version: 1,
    result: "finished",
    roomId,
    status: "active",
    roomRevision: 5,
    progressRevision: 3,
    membershipRevision: 2,
    finishedAt: "2026-08-10T10:30:00Z"
  }).result, "finished");

  assert.throws(() => api.applyResult({
    version: 1,
    result: "applied",
    roomId,
    roomRevision: 4,
    progressRevision: 2,
    kind: "undo_set",
    setId: "s_01_01",
    completedAt: "2026-08-10T10:06:00Z"
  }), /apply result is invalid/);
});

test("realtime event is a bounded invalidation hint, never a workout payload", () => {
  const api = contract();
  assert.deepEqual(JSON.parse(JSON.stringify(api.realtimeEvent({
    version: 1,
    kind: "progress",
    roomId,
    roomRevision: 7
  }))), { version: 1, kind: "progress", roomId, roomRevision: 7 });
  assert.throws(() => api.realtimeEvent({
    version: 1,
    kind: "progress",
    roomId,
    roomRevision: 7,
    weight: 200
  }), /fields are invalid/);
});

File: app/src/test/java/com/example/gymapp/auth/LiveWorkoutContractTest.kt
package com.example.gymapp.auth

import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

class LiveWorkoutContractTest {
    @Test
    fun `gateway request accepts only canonical versioned live actions`() {
        val actions = listOf(
            "live_inbox",
            "live_send_invite",
            "live_respond_invite",
            "live_start",
            "live_snapshot",
            "live_apply",
            "live_finish",
            "live_leave",
            "live_cancel"
        )

        actions.forEach { action ->
            val request = JSONObject(
                liveGatewayRequestJson(action, JSONObject().put("roomId", ROOM_ID))
            )
            assertEquals(
                setOf("version", "action", "payload"),
                request.keys().asSequence().toSet()
            )
            assertEquals(1, request.getInt("version"))
            assertEquals(action, request.getString("action"))
            assertEquals(ROOM_ID, request.getJSONObject("payload").getString("roomId"))
        }
        assertThrows(IllegalArgumentException::class.java) {
            liveGatewayRequestJson("snapshot", JSONObject())
        }
    }

    @Test
    fun `gateway success unwrap is strict and preserves inner contract`() {
        val inner = JSONObject()
            .put("version", 1)
            .put("result", "submitted_or_unavailable")
            .put("roomId", JSONObject.NULL)
            .put("status", JSONObject.NULL)
            .put("roomRevision", JSONObject.NULL)
        val wrapped = JSONObject().put("version", 1).put("result", inner)

        val parsed = parseLiveSendInviteResult(unwrapLiveGatewaySuccess(wrapped.toString()))

        assertEquals("submitted_or_unavailable", parsed.result)
        wrapped.put("trace", "private")
        assertThrows(IllegalArgumentException::class.java) {
            unwrapLiveGatewaySuccess(wrapped.toString())
        }
    }

    @Test
    fun `realtime signal is strict and only acts as an invalidation hint`() {
        val parsed = parseLiveWorkoutRealtimeSignal(
            JSONObject()
                .put("version", 1)
                .put("kind", "progress")
                .put("roomId", ROOM_ID)
                .put("roomRevision", 7)
                .toString()
        )

        assertEquals("progress", parsed.kind)
        assertEquals(7, parsed.roomRevision)

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutRealtimeSignal(
                JSONObject()
                    .put("version", 1)
                    .put("kind", "progress")
                    .put("roomId", ROOM_ID)
                    .put("roomRevision", 7)
                    .put("completedSets", JSONArray())
                    .toString()
            )
        }
    }

    @Test
    fun `inbox parser accepts bounded exact contract`() {
        val response = JSONObject()
            .put("version", 1)
            .put("invitations", JSONArray().put(invitationJson()))
            .put("rooms", JSONArray())
            .toString()

        val parsed = parseLiveWorkoutInbox(response)

        assertEquals(1, parsed.invitations.size)
        assertEquals(ROOM_ID, parsed.invitations.single().roomId)
        assertTrue(parsed.rooms.isEmpty())
    }

    @Test
    fun `inbox parser rejects unknown fields and duplicate rooms`() {
        val withUnknownField = JSONObject()
            .put("version", 1)
            .put("invitations", JSONArray())
            .put("rooms", JSONArray())
            .put("debug", true)

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutInbox(withUnknownField.toString())
        }

        val response = JSONObject()
            .put("version", 1)
            .put("invitations", JSONArray().put(invitationJson()))
            .put("rooms", JSONArray().put(roomJson()))

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutInbox(response.toString())
        }

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutInbox(
                JSONObject()
                    .put("version", 1)
                    .put("invitations", JSONArray())
                    .put("rooms", JSONArray())
                    .toString() + " trailing"
            )
        }
    }

    @Test
    fun `snapshot parser validates canonical set mapping and progress`() {
        val parsed = parseLiveWorkoutSnapshot(snapshotJson().toString())

        assertEquals(setOf("s_01_01"), parsed.plan.setIds)
        assertEquals("s_01_01", parsed.self.progress?.undoableSetId)
        assertEquals("Partner", parsed.peer.profile.displayName)
    }

    @Test
    fun `snapshot parser rejects progress for a set outside canonical plan`() {
        val response = snapshotJson()
        response.getJSONArray("participants")
            .getJSONObject(0)
            .getJSONObject("progress")
            .getJSONArray("completedSets")
            .getJSONObject(0)
            .put("setId", "s_01_02")

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutSnapshot(response.toString())
        }
    }

    @Test
    fun `snapshot parser rejects an undo marker that is not the latest completed set`() {
        val response = snapshotJson()
        val progress = response.getJSONArray("participants")
            .getJSONObject(0)
            .getJSONObject("progress")
        progress.put("undoableSetId", JSONObject.NULL)

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveWorkoutSnapshot(response.toString())
        }
    }

    @Test
    fun `send parser preserves indistinguishable unavailable result`() {
        val parsed = parseLiveSendInviteResult(
            JSONObject()
                .put("version", 1)
                .put("result", "submitted_or_unavailable")
                .put("roomId", JSONObject.NULL)
                .put("status", JSONObject.NULL)
                .put("roomRevision", JSONObject.NULL)
                .toString()
        )

        assertEquals("submitted_or_unavailable", parsed.result)
        assertNull(parsed.roomId)
        assertNull(parsed.status)
        assertNull(parsed.roomRevision)
    }

    @Test
    fun `accept parser prefers active and keeps ready only as response compatibility`() {
        fun response(status: String) = JSONObject()
            .put("version", 1)
            .put("result", "joined")
            .put("roomId", ROOM_ID)
            .put("status", status)
            .put("roomRevision", 5)
            .put("membershipRevision", 2)
            .toString()

        assertEquals("active", parseLiveRespondInviteResult(response("active")).status)
        assertEquals("ready", parseLiveRespondInviteResult(response("ready")).status)
        assertThrows(IllegalArgumentException::class.java) {
            parseLiveRespondInviteResult(response("waiting"))
        }
    }

    @Test
    fun `apply parser requires completion timestamp only for complete`() {
        val complete = parseLiveApplyResult(
            JSONObject()
                .put("version", 1)
                .put("result", "applied")
                .put("roomId", ROOM_ID)
                .put("roomRevision", 4)
                .put("progressRevision", 2)
                .put("kind", "complete_set")
                .put("setId", "s_01_01")
                .put("completedAt", NOW)
                .toString()
        )

        assertTrue(complete is LiveApplyResult.Applied)

        val incoherent = JSONObject()
            .put("version", 1)
            .put("result", "applied")
            .put("roomId", ROOM_ID)
            .put("roomRevision", 4)
            .put("progressRevision", 2)
            .put("kind", "undo_set")
            .put("setId", "s_01_01")
            .put("completedAt", NOW)

        assertThrows(IllegalArgumentException::class.java) {
            parseLiveApplyResult(incoherent.toString())
        }
    }

    @Test
    fun `gateway exception exposes only expected recovery categories`() {
        val conflict = LiveWorkoutGatewayException(409, "conflict", "stale")
        val unavailable = LiveWorkoutGatewayException(404, "resource_unavailable", "gone")

        assertTrue(conflict.isConflict)
        assertFalse(conflict.isResourceUnavailable)
        assertTrue(unavailable.isResourceUnavailable)
        assertFalse(unavailable.isConflict)
    }

    private fun invitationJson(): JSONObject = JSONObject()
        .put("roomId", ROOM_ID)
        .put("status", "waiting")
        .put("roomRevision", 1)
        .put("createdAt", NOW)
        .put("inviteExpiresAt", LATER)
        .put("summary", summaryJson())
        .put("owner", profileJson("p_${"1".repeat(32)}", "Owner"))

    private fun roomJson(): JSONObject = JSONObject()
        .put("roomId", ROOM_ID)
        .put("status", "waiting")
        .put("roomRevision", 1)
        .put("role", "owner")
        .put("memberState", "joined")
        .put("membershipRevision", 1)
        .put("createdAt", NOW)
        .put("startedAt", JSONObject.NULL)
        .put("activeExpiresAt", JSONObject.NULL)
        .put("summary", summaryJson())
        .put("peer", profileJson("p_${"2".repeat(32)}", "Partner"))

    private fun snapshotJson(): JSONObject {
        val complete = JSONObject()
            .put("setId", "s_01_01")
            .put("weight", 80.0)
            .put("reps", 8)
            .put("completedAt", NOW)
        val selfProgress = JSONObject()
            .put("revision", 2)
            .put("completedSets", JSONArray().put(complete))
            .put("undoableSetId", "s_01_01")
            .put("finishedAt", JSONObject.NULL)
        val peerProgress = JSONObject()
            .put("revision", 1)
            .put("completedSets", JSONArray())
            .put("undoableSetId", JSONObject.NULL)
            .put("finishedAt", JSONObject.NULL)
        val participants = JSONArray()
            .put(participantJson(true, "p_${"1".repeat(32)}", "Owner", "owner", selfProgress))
            .put(participantJson(false, "p_${"2".repeat(32)}", "Partner", "participant", peerProgress))
        val plan = JSONObject()
            .put("version", 1)
            .put(
                "exercises",
                JSONArray().put(
                    JSONObject()
                        .put("exerciseId", "e_01")
                        .put("name", "Bench press")
                        .put("catalogKey", "bench_press")
                        .put(
                            "sets",
                            JSONArray().put(
                                JSONObject()
                                    .put("setId", "s_01_01")
                                    .put("weight", 80.0)
                                    .put("reps", 8)
                            )
                        )
                )
            )
        return JSONObject()
            .put("version", 1)
            .put(
                "room",
                JSONObject()
                    .put("roomId", ROOM_ID)
                    .put("status", "active")
                    .put("roomRevision", 3)
                    .put("closeReason", JSONObject.NULL)
                    .put("createdAt", NOW)
                    .put("inviteExpiresAt", LATER)
                    .put("startedAt", NOW)
                    .put("activeExpiresAt", LATER)
                    .put("endedAt", JSONObject.NULL)
                    .put("summary", summaryJson())
            )
            .put("plan", plan)
            .put("participants", participants)
    }

    private fun participantJson(
        isSelf: Boolean,
        profileId: String,
        name: String,
        role: String,
        progress: JSONObject
    ): JSONObject = JSONObject()
        .put("isSelf", isSelf)
        .put("profile", profileJson(profileId, name))
        .put("role", role)
        .put("state", "joined")
        .put("membershipRevision", 1)
        .put("joinedAt", NOW)
        .put("finishedAt", JSONObject.NULL)
        .put("departedAt", JSONObject.NULL)
        .put("progress", progress)

    private fun summaryJson(): JSONObject = JSONObject()
        .put("exerciseCount", 1)
        .put("setCount", 1)
        .put("exerciseNames", JSONArray().put("Bench press"))

    private fun profileJson(id: String, name: String): JSONObject = JSONObject()
        .put("profileId", id)
        .put("displayName", name)

    private companion object {
        const val ROOM_ID = "lr_0123456789abcdef0123456789abcdef"
        const val NOW = "2026-08-10T09:00:00Z"
        const val LATER = "2026-08-10T11:00:00Z"
    }
}
